### ISSUE TITLE
Handling the case when skype is not installed

### ISSUE THREAD

**Issue by SunyataZero** [EmpathyApp/EmpathyApp#42]
Right now we get this message if skype is not installed when clicking on the skype status/call button:

> The address wasn't understood
> 
> Firefox doesn't know how to open this address, because one of the following protocols (skype) isn't associated with any program or is not allowed in this context.
> 
>     You might need to install other software to open this address.

**Comment by prashanthr** [comment]
I've done some work on a project where a similar application was required. We were trying to launch a video chat client from the browser. We used a custom handler "customcall:123456789" and a custom application was able to launch the video call. For this to work, the custom application had to be installed as a pre-requisite. When the application was installed, there was a registry entry written to handle the protocol "customcall". This is similar to when a browser is installed and set as default, it is set to be the primary application to handle "http"/"https" protocols. 

**Background**
The browser needs to talk to the host computer to find an application that can handle the protocol. To do this the registry is checked. Usually when we install a program on the computer, it writes to the registry to handle certain protocols. For example: When Skype is installed, "skype:1234567890" will launch Skype and execute the call if installed automatically. If not installed, the computer doesn't know how to handle it and therefore, it gives us that message. Different browsers deal with missing protocol handlers differently. You may find that Google Chrome handles the error in a different way when compared to Firefox.

**Possible Solutions**
1. Investigate how to handle missing protocol handlers on host computer from the browser
2. If the above is not possible, We could write a custom application whose sole purpose is to act as the middle ware between browser and skype/VoIP application on the client. The application will directly launch Skype (pass-through) if it exists or if it doesn't it will handle it in a user friendly way by directing the user to install Skype/provide them with a message.

**Comment by prashanthr** [comment]
Some useful reference:
https://developer.mozilla.org/en-US/docs/Web-based_protocol_handlers

**Comment by SunyataZero** [comment]
**Notes for a stackoverflow question**
How can i direct the user to an url if an application is missing

Hi all,
i'm working on a web app that to function needs a piece of software (skype) to be installed on the user's computer. There are three cases when the user clicks a button:
1. the software is installed and started on the users computer
2. the software is installed but not yet started
3. the software has not been installed

Case 1:
works well

Case 2:
skype is launched but the user specified is not called

Case 3:
Nothing happens

**Comment by SunyataZero** [comment]
@prashanthr Hi, thank you for the notes you've written, interesting article for me on mozdev. I'm looking at this now but have not come far yet, but was thinking that maybe the problem is not there for Windows, could you try it on a Windows computer? (i don't have one available myself, sorry). If we don't see it on windows perhaps we can close this issue since most people have this OS

**Comment by prashanthr** [comment]
@SunyataZero Yes, I will look into this. I'll test a few cases and report my findings.

**Comment by SunyataZero** [comment]
Thank you. I tried this code (inserted with a new shortcode on the same page as the skype button) but it did nothing unforturnately:
```
function ea_home_shortcode() {
    ob_start(); //++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
    ?>

    <script>
        jQuery(function() {
            navigator.registerProtocolHandler("skype", "http://www.skype.com", "Skype");
/*
        navigator.registerProtocolHandler("skype",
                                      "http://www.google.co.uk/?uri=%s",
                                      "Burger handler");
                                      */
        });
    </script>

    <?php
    $ob_content = ob_get_contents(); //+++++++++++++++++++++++++++++++++++++++++
    ob_end_clean();
    return $ob_content;
}
add_shortcode('ea_home', 'ea_home_shortcode');
```

**Comment by prashanthr** [comment]
Here is how it looks to a user from a Windows 8 Computer

When the skype icon is clicked from a Firefox browser, the following dialog box is shown
![firefoxhandling](https://cloud.githubusercontent.com/assets/1409501/5952198/d7d93784-a737-11e4-9503-d3cc5eca7283.JPG)

In Google Chrome, it is handled in the following way
![googlechromepermission](https://cloud.githubusercontent.com/assets/1409501/5952257/cef49bda-a738-11e4-8400-dd3b0e0165ee.JPG)


Both browsers use "Pick An App" to handle the request. "Pick An App" is a Windows 8 default handler that opens apps after checking the registry for a way to handle "skype:" protocols. You will then see a pop up as shown below and the user can pick Skype if it is installed. If not, the user will have to download an app from the Windows store. 
![win-8-pick-an-app](https://cloud.githubusercontent.com/assets/1409501/5952255/bf2ae6aa-a738-11e4-9a57-11c91664301a.jpg)

**Comment by SunyataZero** [comment]
Thank you for this, very useful seeing the actual screenshots, i try to do
this as well sometimes. Ok, so the skype alternatives are shown even if
skype is not installed, then maybe solving this issue is not necessary for
version 1.0? perhaps for 1.1 or something? i'm creating a new milestone
called 1.1 now where i'll add some features we would like to have but that
we may not have time for during the coming weeks

On Thu, Jan 29, 2015 at 5:00 AM, Prashanth Rajaram <notifications@github.com
> wrote:

> Here is how it looks to a user from a Windows 8 Computer
>
> When the skype icon is clicked from a Firefox browser, the following
> dialog box is shown
> [image: firefoxhandling]
> <https://cloud.githubusercontent.com/assets/1409501/5952198/d7d93784-a737-11e4-9503-d3cc5eca7283.JPG>
>
> In Google Chrome, it is handled in the following way
> [image: googlechromepermission]
> <https://cloud.githubusercontent.com/assets/1409501/5952257/cef49bda-a738-11e4-8400-dd3b0e0165ee.JPG>
>
> Both browsers use "Pick An App" to handle the request. "Pick An App" is a
> Windows 8 default handler that opens apps after checking the registry for a
> way to handle "skype:" protocols. You will then see a pop up as shown below
> and the user can pick Skype if it is installed. If not, the user will have
> to download an app from the Windows store.
> [image: win-8-pick-an-app]
> <https://cloud.githubusercontent.com/assets/1409501/5952255/bf2ae6aa-a738-11e4-9a57-11c91664301a.jpg>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/EmpathyApp/EmpathyApp/issues/42#issuecomment-71965514>
> .
>



-- 
https://github.com/EmpathyApp/EmpathyApp/
<http://www.svenskafreds.se/fredskalender-for-goteborg>
http://ihavearrived.org/

**Comment by SunyataZero** [comment]
@prashanthr Unassigning you from this issue, hope this is ok? Because @Peijman wants us to focus on the BETA version right now and i don't think we need this

**Comment by prashanthr** [comment]
@SunyataZero No problem! I completely understand. We can investigate this when we deal with this issue when it becomes more critical/relevant.
